google-cloud-bigtable/src/test/java/com/google/cloud/bigtable/test_helpers/env/CloudEnv.java

```diff
@@ -153,4 +193,53 @@ private static String getRequiredProperty(String prop) {
     }
     return value;
   }
+
+  /**
+   * Captures the request attributes "Grpc.TRANSPORT_ATTR_REMOTE_ADDR" when connection is established.
+   * This is useful for DirectPath testing and debugging.
+   * {@link ClientCall#getAttributes()}
+   */
+  private ClientInterceptor recordClientCallInterceptor(
+      final AtomicReference<ClientCall<?, ?>> clientCallCapture) {
+    return new ClientInterceptor() {
+      @Override
+      public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
+          MethodDescriptor<ReqT, RespT> method, CallOptions callOptions, Channel next) {
+        ClientCall<ReqT, RespT> clientCall = next.newCall(method,callOptions);
+        clientCallCapture.set(clientCall);
+        return new SimpleForwardingClientCall<ReqT, RespT>(clientCall) {
+          @Override
+          public void start(Listener<RespT> responseListener, Metadata headers) {
+            super.start(new SimpleForwardingClientCallListener<RespT>(responseListener) {
+              @Override
+              public void onHeaders(Metadata headers) {
+                // Check peer IP after connection is established.
+                SocketAddress remoteAddr = clientCallCapture.get().getAttributes().get(Grpc.TRANSPORT_ATTR_REMOTE_ADDR);
+                if (!verifyRemoteAddress(remoteAddr)) {
+                  throw new RuntimeException(String.format(
```

Comment:
Yes, it will throw `java.util.concurrent.ExecutionException: com.google.api.gax.rpc.CancelledException: io.grpc.StatusRuntimeException: CANCELLED: Failed to read headers`

Reply:
Good to know that its safe, but the test failure will be hard for someone to track down.
Can you make the message more explicit, something along the lines of:

"Synthetically aborting the current request because it did not adhere to the test environment's requirement for DirectPath. Expected test to access DirectPath via %s, but RPC was destined for %s"